src/Microsoft.AspNet.Hosting/Startup/FailSafeStartupLoader.cs

```diff
@@ -0,0 +1,168 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+using Microsoft.Dnx.Runtime;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.Logging;
+
+namespace Microsoft.AspNet.Hosting.Startup
+{
+    /// <summary>
+    /// Catches common startup exceptions and shows an error page.
+    /// </summary>
+    public class FailSafeStartupLoader : IStartupLoader
```

Comment:
Don't implement as an IStartupLoader. Instead put `try {} catch {}` in `HostingEngine.BuildApplication` catch should determine if the error should be rethrown or displayed. If displayed it should return a simple delegate - doesn't need to follow Startup-class logic paths.

This also means the call to EnsureApplicationServices in HostingEngine.Start should move down into the BuildApplication method.

TBH, the application-services assigning and telemetry-providing delegate wrapper should be also be moved into the BuildApplication success path instead of the engine Start method.